src/components/tl/ucp/reduce/reduce_srg_knomial.c

```diff
@@ -0,0 +1,334 @@
+    if (schedule->scratch_mc_header) {
+        ucc_mc_free(schedule->scratch_mc_header);
+    }
+    status = ucc_schedule_pipelined_finalize(task);
+    ucc_tl_ucp_put_schedule(&schedule->super.super);
+    return status;
+}
+
+ucc_status_t ucc_tl_ucp_reduce_srg_knomial_start(ucc_coll_task_t *task)
+{
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(task, "ucp_reduce_srg_kn_start", 0);
+    return ucc_schedule_pipelined_post(task);
+}
+
+static void
+ucc_tl_ucp_reduce_srg_knomial_get_pipeline_params(ucc_tl_ucp_team_t *team,
+                                                  ucc_memory_type_t mtype,
+                                                  ucc_pipeline_params_t *pp)
+{
+    ucc_tl_ucp_lib_config_t *cfg = &team->cfg;
+    ucc_mc_attr_t mc_attr;
+
+    if (!ucc_pipeline_params_is_auto(&cfg->reduce_srg_kn_pipeline)) {
+        *pp = cfg->reduce_srg_kn_pipeline;
+        return;
+    }
+
+    if (mtype == UCC_MEMORY_TYPE_CUDA) {
+        mc_attr.field_mask = UCC_MC_ATTR_FIELD_FAST_ALLOC_SIZE;
+        ucc_mc_get_attr(&mc_attr, UCC_MEMORY_TYPE_CUDA);
+        pp->threshold = mc_attr.fast_alloc_size;
+        pp->n_frags   = 2;
+        pp->frag_size = mc_attr.fast_alloc_size;
+        pp->order     = UCC_PIPELINE_PARALLEL;
+        pp->pdepth    = 2;
+    } else {
+        pp->threshold = SIZE_MAX;
+        pp->n_frags   = 0;
+        pp->frag_size = 0;
+        pp->pdepth    = 1;
+        pp->order     = UCC_PIPELINE_PARALLEL;
+    }
+}
+
+ucc_status_t ucc_tl_ucp_reduce_srg_knomial_init(ucc_base_coll_args_t *coll_args,
+                                                ucc_base_team_t *team,
+                                                ucc_coll_task_t **task_h)
+{
+    ucc_tl_ucp_team_t     *tl_team = ucc_derived_of(team, ucc_tl_ucp_team_t);
+    ucc_coll_args_t       *args    = &coll_args->args;
+    ucc_rank_t             trank   = UCC_TL_TEAM_RANK(tl_team);
+    int                    n_frags, pipeline_depth;
+    ucc_tl_ucp_schedule_t *schedule;
+    ucc_status_t           st;
+    ucc_base_coll_args_t   bargs;
+    size_t                 max_frag_count, dt_size, count;
+    ucc_pipeline_params_t  pipeline_params;
+    ucc_datatype_t         dt;
+    ucc_memory_type_t      mt;
+
+    st  = ucc_tl_ucp_get_schedule(tl_team, coll_args, &schedule);
+    if (ucc_unlikely(UCC_OK != st)) {
+        goto err_out;
+    }
+
+    schedule->scratch_mc_header = NULL;
+    if (UCC_IS_ROOT(*args, trank)) {
+        count   = args->dst.info.count;
+        dt      = args->dst.info.datatype;
+        mt      = args->dst.info.mem_type;
+        dt_size = ucc_dt_size(dt);
+    } else  {
+        count   = args->src.info.count;
+        dt      = args->src.info.datatype;
+        mt      = args->src.info.mem_type;
+        dt_size = ucc_dt_size(dt);
+    }
+
+    if (!UCC_IS_ROOT(*args, trank) || UCC_IS_INPLACE(*args)) {
+        st = ucc_mc_alloc(&schedule->scratch_mc_header, count * dt_size, mt);
```

Comment:
Can you avoid the scratch buffer allocation? For example, in the first call of this collective algorithm, allocate a large enough buffer for this team, and reuse this scratch buffer for the future calls into this collective again.

Reply:
No. Scratch space is shared between TLs and CLs and there is also fast scratch allocation pool. If some team doesn't return scratch back to that pool we will run out of fast allocation cache pretty quickly.